src/microbe_stage/organelle_unlocks/OrganelleUnlockConstraints.cs

```diff
@@ -0,0 +1,108 @@
+/// </summary>
+public class OrganelleUnlockConstraints
+{
+    /// <summary>
+    ///   The number of microbes engulfed by the player is above the specified amount.
+    /// </summary>
+    public int? EngulfedMicrobesAbove { get; set; }
+
+    /// <summary>
+    ///   The number of times the player has died is above a specified amount.
+    /// </summary>
+    public int? PlayerDeathsAbove { get; set; }
+
+    /// <summary>
+    ///   The player species produces more than a specified amount of APT.
+    /// </summary>
+    public float? AptAbove { get; set; }
+
+    /// <summary>
+    ///   The player species produces a specified amount more APT than it uses.
+    /// </summary>
+    public float? ExcessAptAbove { get; set; }
+
+    /// <summary>
+    ///   The player species has a base speed that is slower than a certain speed.
+    /// </summary>
+    public float? BaseSpeedBelow { get; set; }
+
+    /// <summary>
+    ///   The player species has had a specific organelle for some number of generations.
+    /// </summary>
+    public ReproducedWith? ReproducedWith { get; set; }
+
+    /// <summary>
+    ///   The player has reproduced in a specific named biome.
+    /// </summary>
+    public string? ReproduceInBiome { get; set; }
+
+    /// <summary>
+    ///   The player's current biome has a certain amount of a compound.
+    /// </summary>
+    public BiomeCompound? BiomeCompound { get; set; }
+
+    /// <summary>
+    ///   Are all of the specified unlock constraints satisfied?
+    /// </summary>
+    public bool Satisfied(GameWorld world)
+    {
+        if (world.PlayerSpecies is not MicrobeSpecies playerSpecies)
+            return false;
+
+        var energyBalance = ProcessSystem.ComputeEnergyBalance(playerSpecies.Organelles, world.Map.CurrentPatch!.Biome,
+            playerSpecies.MembraneType, true, world.WorldSettings, CompoundAmountType.Current);
+
+        return UnlockConstraints().All(constraint => constraint.Satisfied(world, energyBalance));
+    }
+
+    /// <summary>
+    ///   Display the requirements as a string.
+    /// </summary>
+    public string Tooltip(GameWorld world)
+    {
+        if (world.PlayerSpecies is not MicrobeSpecies playerSpecies)
+            return "In the microbe stage";
+
+        var energyBalance = ProcessSystem.ComputeEnergyBalance(playerSpecies.Organelles, world.Map.CurrentPatch!.Biome,
+            playerSpecies.MembraneType, true, world.WorldSettings, CompoundAmountType.Current);
+
+        return UnlockConstraints().Select(constraint => constraint.Tooltip(world, energyBalance)).Join();
+    }
+
+    /// <summary>
+    ///   An iterable over all of the unlock constraints that have been set.
+    /// </summary>
+    private IEnumerable<IUnlockConstraint> UnlockConstraints()
+    {
+        if (EngulfedMicrobesAbove is not null)
+            yield return new EngulfedMicrobesAbove { Microbes = EngulfedMicrobesAbove.Value };
```

Comment:
Unless I misunderstood C# boxing rules, the use of structs is negated here with basically mandatory boxing to transfer the data through an enumerator out of this method. Also this creating of temporary objects to check the conditions isn't exactly my preferred method I'd design this. Instead even when loading the data from JSON the actual class types that would do the checking can be loaded directly. For example see how organelle component factories are specified for the OrganelleDefinition directly in the JSON.